### DISCUSSION TITLE
`composer.lock`’s mtime more recent than `vendor`’s

### DISCUSSION THREAD

**Discussion by MatTheCat** [composer/composer#11405] [Q&A / Support] [answered]
Hi,

AFAIU Composer `touch` the `vendor` directory if any installation was performed, and this happens after the lock file is written. Thus, I have a Makefile rule whose target is `vendor` and only prerequisite is `composer.lock`.

But from time to time, `make` keeps executing the recipe even if nothing changed. The last time I checked my `vendor`’s mtime was two seconds older than `composer.lock`’s. How could this happen?

I run Arch Linux and I can access these two files with PhpStorm on my host. These are written in a docker container; could it be the cause?

**Comment by Seldaek** [comment]
I really don't see how this could happen as it does `file_put_contents` on lock file and very shortly after does `touch` on vendor dir. Could be that docker messes with the clock, I'm not familiar enough with it to say.

**Reply by stof** [reply]
could it be the same issue than https://gist.github.com/Seldaek/a5cf4a5551139bc41fb130fa16406290, for which Packagist has a workaround in its endpoint listing updated packages ? https://github.com/composer/packagist/blob/0d63897e12bfcc25049f63659626af42ff5277ed/src/Controller/PackageController.php#L181-L184

**Reply by MatTheCat** [reply]
It seems I can consistently reproduce by removing then requiring the same package. I did it outside of Docker using my terminal only to make sure my IDE wasn’t messing with mtimes.

```shell
$ composer remove twig/intl-extra
…

$ find composer.lock vendor -maxdepth 0 -printf '%t %p\n'
Wed Mar 29 18:34:49.6940273060 2023 composer.lock
Wed Mar 29 18:34:49.6940273060 2023 vendor

$ composer require twig/intl-extra
…

$ find composer.lock vendor -maxdepth 0 -printf '%t %p\n'
Wed Mar 29 18:39:49.6583335460 2023 composer.lock
Wed Mar 29 18:39:48.6349968790 2023 vendor
```

**Reply by stof** [reply]
The issue might be the version guessing done in `composer require` where it first runs the update with a `*` constraint and then alter the composer.json with the guessed constraint. At that point, it probably modify the composer.lock file again to update the hash in it, but doing that _after_ touching the vendor folder.

**Reply by Seldaek** [reply]
Ah yes that's exactly what's going on, we should probably ensure we keep the filemtime when patching the lock file like that.

**Reply by MatTheCat** [reply]
I can confirm running `composer require "twig/intl-extra ^3.5"` results in `composer.lock` and `vendor` having the same mtime.

**Reply by stof** [reply]
@Seldaek either preserving the filemtime or touching the vendor folder again.

**Reply by Seldaek** [reply, marked as answer]
Ok should be fixed by https://github.com/composer/composer/commit/2d2d22d0ecfa9148f15ce13969fd614417d26ab1 - you can check with `composer self-update --snapshot` if it resolves the issue for you.

**Reply by MatTheCat** [reply]
Just tested and issue is still there :confused: Shouldn't the commit touch the vendor directory instead of the lock file?

**Reply by stof** [reply]
Well, the commit touches the lock file to preserve its mtime

**Reply by MatTheCat** [reply]
Oh okay missed `touch`’s second argument.

My issue was that `composer self-update --snapshot` first updated to 595559f68d8c01715a2be33a54247898f9fbc34e :thinking: 
After running once more I got the correct version and I can confirm the fix is working :tada: 

Thanks!

**Reply by stof** [reply]
Well, maybe your first self-update ran before the generation of the new snapshot phar was over
